cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -0,0 +1,218 @@
+import com.google.inject.name.Named;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Process metrics by consuming metrics being published to TMS.
+ */
+public class MessagingMetricsProcessorService extends AbstractMetricsProcessorService {
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingMetricsProcessorService.class);
+
+  private static final long INITIAL_LAST_MESSAGE_ID = -1L;
+
+  private final TopicId metricsTopic;
+  // TODO unused partitions
+  private final Set<Integer> partitions;
+  private final MessagingService messagingService;
+  private final DatumReader<MetricValues> recordReader;
+  private final Schema recordSchema;
+  private final MetricStore metricStore;
+  private Map<String, String> metricsContextMap;
+  private final int persistThreshold;
+  private final AtomicInteger messageCount;
+
+  private long lastLoggedMillis;
+  private long recordsProcessed;
+  private AtomicLong lastMessageId;
+
+  @Inject
+  public MessagingMetricsProcessorService(MetricDatasetFactory metricDatasetFactory,
+                                          @Named(Constants.Metrics.MESSAGING_TOPIC_PREFIX) String topicPrefix,
+                                          @Assisted Set<Integer> partitions,
+                                          MessagingService messagingService,
+                                          SchemaGenerator schemaGenerator,
+                                          DatumReaderFactory readerFactory,
+                                          MetricStore metricStore,
+                                          @Named(Constants.Metrics.MESSAGING_FETCHER_PERSIST_THRESHOLD)
+                                          int persistThreshold) {
+    super(metricDatasetFactory);
+    this.metricsTopic = NamespaceId.SYSTEM.topic(topicPrefix);
+    this.partitions = partitions;
+    this.messagingService = messagingService;
+    try {
+      this.recordSchema = schemaGenerator.generate(MetricValues.class);
+      this.recordReader = readerFactory.create(TypeToken.of(MetricValues.class), recordSchema);
+    } catch (UnsupportedTypeException e) {
+      throw Throwables.propagate(e);
+    }
+    this.metricStore = metricStore;
+    this.persistThreshold = persistThreshold;
+    this.messageCount = new AtomicInteger();
+    this.lastMessageId = new AtomicLong(INITIAL_LAST_MESSAGE_ID);
+    this.metricsContextMap = Collections.<String, String>emptyMap();
+  }
+
+  @Override
+  public void setMetricsContext(MetricsContext metricsContext) {
+    this.metricsContext = metricsContext;
+    this.metricsContextMap = metricsContext == null ? Collections.<String, String>emptyMap() : metricsContext.getTags();
```

Comment:
who passes a null metricsContext? If it can be null, it should be annotated as `@Nullable`